sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,726 @@
+	//
+	// If we load a bucket that has sub-buckets but is non-empty,
+	// delete all of the sub-buckets.  (The current implementation recurses
+	// so there may be more than one level of sub-buckets created by shardBucket.)
+
+	// Visit the buckets using inorder traversal, parents before children.
+	sort.Strings(keys)
+
+	nonemptyParent := "NOT_A_PREFIX"
+
+	s.bucketsCacheLock.Lock()
+	defer s.bucketsCacheLock.Unlock()
+	for _, key := range keys {
+		if strings.HasPrefix(key, nonemptyParent) {
+			s.Logger.Warn("detected shadowed bucket, removing", "key", key, "parent", nonemptyParent)
+			s.BucketStorageView.Delete(ctx, key)
+			continue
+		}
+
+		// Read from the underlying view
+		storageEntry, err := s.BucketStorageView.Get(ctx, key)
+		if err != nil {
+			return errwrap.Wrapf("failed to read packed storage entry: {{err}}", err)
+		}
+		if storageEntry == nil {
+			return fmt.Errorf("no data found at bucket %s", key)
+		}
+		bucket, err := s.DecodeBucket(storageEntry)
+		if err != nil {
+			return err
+		}
+		if !bucket.HasShards {
+			nonemptyParent = key
+		}
+
+		s.bucketsCache.Insert(s.GetCacheKey(bucket.Key), bucket)
+	}
+
+	s.addAllBaseBucketsToCache()
+	return nil
+}
+
+// Ensure all base buckets are present on the cache, even if they
+// are not present in storage.  This simplifies acquiring locks for
+// multiple buckets because every necessary bucket is already present
+// in bucketsCache. We don't have to worry about adding them on-demand
+// in the middle of that process, and if any object is added to these
+// placeholders they'll get saved to stoarge.
+func (s *StoragePackerV2) addAllBaseBucketsToCache() {
+	for _, bucketKey := range s.getAllBaseBucketKeys() {
+		if _, present := s.bucketsCache.Get(bucketKey); !present {
+			lock := locksutil.LockForKey(s.storageLocks, bucketKey)
+			s.bucketsCache.Insert(bucketKey,
+				&LockedBucket{
+					LockEntry: lock,
+					Bucket: &Bucket{
+						Key:       bucketKey,
+						ItemMap:   make(map[string][]byte),
+						HasShards: false,
+					},
+				})
+		}
+	}
+}
+
+func (s *StoragePackerV2) BucketKeys(ctx context.Context) ([]string, error) {
+	var retErr error
+	s.prewarmCache.Do(func() {
+		diskBuckets, err := logical.CollectKeys(ctx, s.BucketStorageView)
+		if err != nil {
+			retErr = err
+			return
+		}
+		retErr = s.preloadBucketsFromDisk(ctx, diskBuckets)
+	})
+	if retErr != nil {
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
+
+	return ret, nil
+}
+
+// DecodeBucket parses a bucket from storage, but doesn't add it to the cache.
+// (We may want to read a v1 bucket and upgrade it?)
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	if storageEntry == nil || storageEntry.Value == nil {
+		return nil, errors.New("decoding nil storageEntry")
```

Comment:
```suggestion
		return nil, errors.New("decoding nil storage entry")
```

Reply:
I prefer the actual name of the argument here.